lib/caqti_request.mli

```diff
@@ -28,38 +28,69 @@
     query. *)
 
 (** {2 Primitives} *)
+type query = Caqti_sql.query
 
-type query =
-  | L of string  (** Literal code. May contain incomplete fragments. *)
-  | P of int     (** [P i] refers to parameter number [i], counting from 0. *)
-  | S of query list (** [S frags] is the concatenation of [frags]. *)
-(** A representation of a query string to send to a database, abstracting over
-    parameter references and providing nested concatenation to simplify
-    generation.  For databases which only support linear parameters (typically
-    denoted "[?]"), the driver will reshuffle, elide, and duplicate parameters
-    as needed.  *)
-
-type ('a, 'b, +'m) t constraint 'm = [< `Zero | `One | `Many]
+type counit
+(** Empty type *)
+
+val counit : unit -> counit
+
+type ('params, 'input, 'output, +'mult) t4 constraint 'mult = [< `Zero | `One | `Many | `Many_in]
+(** A request specification embedding a query generator, parameter encoder, and
+    row decoder.
+    - ['params] is the type of the expected parameter bundle.
+    - ['input] is the type of an input row.
+    - ['output] is the type of a returned row.
+    - ['mult] is the possible multiplicities of returned rows. *)
+
+type ('params, 'output, +'mult) t = ('params, counit, 'output, 'mult) t4
 (** A request specification embedding a query generator, parameter encoder, and
     row decoder.
-    - ['a] is the type of the expected parameter bundle.
-    - ['b] is the type of a returned row.
-    - ['m] is the possible multiplicities of returned rows. *)
+    - ['params] is the type of the expected parameter bundle.
+    - ['output] is the type of a returned row.
+    - ['multiplicity] is the possible multiplicities of returned rows. *)
+
+val create_full :
+  ?oneshot: bool ->
+  'params Caqti_type.t ->
+  'input Caqti_type.t ->
+  'output Caqti_type.t ->
+  'mult Caqti_mult.t ->
+  (Caqti_driver_info.t -> query) ->
+  ('params, 'input, 'output, 'mult) t4
+(** [create_full params_type input_type output_type row_mult f] is a request which takes
+    parameters of type [params_type], inputs rows of type [input_type] and returns rows of
+    type [output_type] with multiplicity [row_mult], and which sends query strings generated
+    from the query [f di], where [di] is the {!Caqti_driver_info.t} of the target driver.
+    The driver is responsible for turning parameter references into a form accepted by the
+    database, while other differences must be handled by [f]. *)
```

Comment:
Yes, and we'll need a corresponding `call_full` in `Caqti_connection_sig`, where `call` becomes a special case passing an empty stream.  That will avoid the problematic `collapse` function.